packages/x-charts/src/internals/useAnimate.ts

```diff
@@ -0,0 +1,116 @@
+'use client';
+import useId from '@mui/utils/useId';
+import * as React from 'react';
+import { interrupt, Transition } from '@mui/x-charts-vendor/d3-transition';
+import { select } from '@mui/x-charts-vendor/d3-selection';
+import { ANIMATION_DURATION_MS, ANIMATION_TIMING_FUNCTION_JS } from '../constants';
+
+export function useAnimate<Props, Elem extends Element>(
+  props: Props,
+  {
+    createInterpolator,
+    applyProps,
+    skip,
+  }: {
+    createInterpolator: (lastProps: Props, newProps: Props) => (t: number) => Props;
+    applyProps: (element: Elem, props: Props) => void;
+    skip?: boolean;
+  },
+) {
+  const transitionName = `MuiUseAnimate-${useId()}`;
+  /* A transition is interrupted when:
+   * T1. The transitioned element changes;
+   * T2. The transitioned element is unmounted;
+   * T3. The component calling this hook is unmounted. */
+  const lastInterpolatedProps = React.useRef(props);
+  const transitionRef = React.useRef<Transition<Elem, unknown, null, undefined>>(null);
+  const elementRef = React.useRef<Elem>(null);
+  const elementUnmounted = React.useRef(false);
+
+  React.useLayoutEffect(() => {
+    return () => {
+      /* T3. Stop the transition if the component that calls this hook is unmounted. */
+      const lastElement = elementRef.current;
+
+      if (lastElement) {
+        interrupt(lastElement, transitionName);
+      }
+    };
+  }, [transitionName]);
+
+  /* T2. Interrupt the transition if the element is unmounted.
+   *
+   * `elementUnmounted` is when to true when a `setRef` is called with a non-null element. This is needed because
+   * `setRef` can be called because the component using this hook re-renders, i.e., it isn't guaranteed to be called only
+   *  when the underlying element changes.
+   *  When `elementUnmounted` is true, it means `setRef` wasn't called with an element, so we must interrupt the
+   *  transition.
+   *
+   * This runs on every render, so it must be light. */
+  React.useLayoutEffect(() => {
```

Comment:
I'm surprised the server does nto complain that `useLayoutEffect` does not exist on server

Reply:
That's a good point. I have no idea why I'm not seeing a warning 😅 
Could it be because the playground is running on NextJS pages router? 

Nevertheless, I think we might have to use an "isomorphic" layout effect, as none of [these options](https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85) apply here. It would be unfortunate to introduce a component just to support SSR.

Reply:
Done 👍